sqlglot/optimizer/qualify_columns.py

```diff
@@ -380,18 +410,105 @@ def _qualify_columns(scope: Scope, resolver: Resolver) -> None:
                     column.set("table", column_table)
 
 
+def _expand_struct_stars(
+    expression: exp.Dot,
+    next_alias_name: t.Callable[[], str],
+) -> t.List[exp.Alias]:
+    """[BigQuery] Expand/Flatten foo.bar.* where bar is a struct column"""
+
+    dot_column = t.cast(exp.Column, expression.find(exp.Column))
+    if not dot_column.is_type(exp.DataType.Type.STRUCT):
+        return []
+
+    # All nested struct values are ColumnDefs, so normalize the first exp.Column in one
+    dot_column = dot_column.copy()
```

Comment:
why does it need to be copied?

Reply:
Let's say we attempt to expand `tbl.col.*`; this expression when entering `_expand_struct_star()` is of type `exp.Dot(....(exp.Column)), exp.Star)`, where the column is the top level struct  i.e `exp.Column(this=col, table=tbl)`.  However, the nested (and annotated) fields of `col` are of type `exp.ColumnDef`, so during the first steps of the traversal we'd have 2 types of expressions to handle.

To canonicalize the procedure, we transform that initial column into an `exp.ColumnDef` as well, but since our expansion can fail e.g. because of unnamed fields we also want to preserve the original expression `tbl.col.*` to fallback to, thus the copy.